arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/CircuitBreaker.kt

```diff
@@ -7,8 +7,121 @@ import arrow.fx.coroutines.CircuitBreaker.State.HalfOpen
+ *    - All requests short-circuit/fail-fast with `ExecutionRejected`
+ *    - If a request is made after the configured [resetTimeout] passes, the [CircuitBreaker] is tripped into the a [HalfOpen] state, allowing one request to go through as a test.
+ *
+ *  3. [HalfOpen]: The [CircuitBreaker] is in this state while it's allowing a request to go through, as as a `test request`
+ *    - All other requests made while `test request` is still running will short-circuit/fail-fast.
+ *    - If the `test request` succeeds then the [CircuitBreaker] is tripped back into [Closed], with the resetTimeout and the failures count also reset to initial values.
+ *    - If the `test request` fails, then the [CircuitBreaker] is tripped back into [Open], the [resetTimeout] is multiplied by the [exponentialBackoffFactor], up to the configured [maxResetTimeout].
+ *
+ * Let's say we'd want to create a [CircuitBreaker] that only allows us to call a remote service twice,
+ * and then whenever more than two requests fail with an exception, the circuit breaker starts short-circuiting failing fast.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.Either
+ * import arrow.core.flatten
+ * import arrow.fx.coroutines.CircuitBreaker
+ * import kotlin.time.Duration
+ * import kotlin.time.ExperimentalTime
+ * import kotlinx.coroutines.delay
+ *
+ * @ExperimentalTime
+ * suspend fun main(): Unit {
+ * //sampleStart
+ *   val circuitBreaker = CircuitBreaker.of(
+ *     maxFailures = 2,
+ *     resetTimeout = Duration.seconds(2),
+ *     exponentialBackoffFactor = 1.2,
+ *     maxResetTimeout = Duration.seconds(60),
+ *   )
+ *   circuitBreaker.protectOrThrow { "I am in Closed: ${circuitBreaker.state()}" }.also(::println)
+ *
+ *   println("Service getting overloaded . . .")
+ *
+ *   Either.catch { circuitBreaker.protectOrThrow { throw RuntimeException("Service overloaded") } }.also(::println)
+ *   Either.catch { circuitBreaker.protectOrThrow { throw RuntimeException("Service overloaded") } }.also(::println)
+ *   circuitBreaker.protectEither { }.also { println("I am Open and short-circuit with ${it}. ${circuitBreaker.state()}") }
+ *
+ *   println("Service recovering . . .").also { delay(2000) }
+ *
+ *   circuitBreaker.protectOrThrow { "I am running test-request in HalfOpen: ${circuitBreaker.state()}" }.also(::println)
+ *   println("I am back to normal state closed ${circuitBreaker.state()}")
+ * //sampleEnd
+ * }
+ * ```
+ *
+ * A common pattern to make fault-tolerant/resilient systems this can be combined with a backing-off policy retry policy to guarantee not overloading the resource
+ * but also not the client that is interacting with the resource.
+ * Below you can see how the simple `retry` function will result in `Either.Left<CircuitBreaker.RejectedExecution>`,
+ * but when we combine it with another schedule, it will always call the `CircuitBreaker` on times that it could've gone in [HalfOpen] state.
```

Comment:
```suggestion
 * but when we combine it with another schedule, it will always call the `CircuitBreaker` on times that it could've entered the [HalfOpen] state.
```